diffrax/integrate.py

```diff
@@ -142,34 +218,107 @@ def body_fun(state, inplace):
-                else:
-                    return x
+                def _body_fun(_val):
+                    y_prev, _, dense_info, _, _, _, step = _val
+                    is_vf_wrapper = lambda x: isinstance(x, VectorFieldWrapper)
+
+                    def _apply_history(x):
+                        if is_vf_wrapper(x):
+                            vector_field = _HistoryVectorField(
+                                x.vector_field,
+                                t0,
+                                state.tprev,
+                                state.tnext,
+                                dense_info,
+                                dense_interp,
+                                solver.interpolation_cls,
+                                y0_history,
+                                delays.delays,
+                            )
+                            return VectorFieldWrapper(vector_field)
+                        else:
+                            return x
+
+                    terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
+                    (y, y_error, dense_info, solver_state, solver_result) = solver.step(
+                        terms_,
+                        state.tprev,
+                        state.tnext,
+                        state.y,
+                        args,
+                        state.solver_state,
+                        state.made_jump,
+                    )
 
-            terms_ = jtu.tree_map(_apply_history, terms, is_leaf=is_vf_wrapper)
-            # TODO: write down implicit problem wrt dense_info, using `terms_`
-            (y, y_error, dense_info, solver_state, solver_result) = terms_  # ...
+                    _pred = (((y - y_prev) / y) > delays.eps).any()
+                    _pred = _pred & (step < 10)
+                    return (
+                        y,
+                        y_error,
+                        dense_info,
+                        solver_state,
+                        solver_result,
+                        _pred,
+                        step + 1,
+                    )
+
+                _init_val = (
+                    state.y,
+                    0,
+                    dict(
+                        zip(
+                            state.dense_infos.keys(),
+                            [
+                                state.dense_infos[key][state.dense_save_index - 1]
+                                for key in state.dense_infos.keys()
+                            ],
+                        )
+                    ),
+                    0,
+                    0,
+                    True,
+                    0,
+                )
+                (
+                    y,
+                    y_error,
+                    dense_info,
+                    solver_state,
+                    solver_result,
+                    _,
+                    _,
+                ) = lax.while_loop(_cond_fun, _body_fun, _init_val)
```

Comment:
By the way, it's not clear to me what the best technique for performing this implicit solve really is. At the moment we're initialising `dense_info` according to whatever it was on the last step, and then performing a fixed-point iteration. Perhaps we should be extrapolating to get a predicted value for `dense_info` on this step, and then using a Newton method? (For example.)

Do you know what the preferred way of solving this implicit problem is?

Reply:
This seems to be the best way to do the implicit solve but it might not work 100% of the time (cf [page 10](http://www.cs.toronto.edu/pub/reports/na/hzpEnrightNA09Preprint.pdf) )